src/api/exchange/ExchangeApi.ts

```diff
@@ -101,21 +115,175 @@ export interface Order {
+
+    return tradeEvents.filter(event => !event['removed']).map(this.parseTradeEvent)
+  }
+
+  public async subscribeToTradeEvent(params: SubscriptionParams): Promise<() => void> {
+    const { userAddress, networkId, callback } = params
+
+    // Remove any active subscription
+    this.unsubscribeToTradeEvent()
+
+    logDebug(`[ExchangeApiImpl] subscribing to trade events for address ${userAddress} and networkId ${networkId}`)
+
+    const subscription = await this.getTradeSubscription(params)
+    // Two ways of subscribing
+
+    // 1: with error handling
+    subscription.subscribe((error, event) => {
+      if (error) {
+        console.error(
+          `[ExchangeApiImpl] Failed to receive Trade event for address ${userAddress} on network ${networkId}: ${error}`,
+        )
+      } else {
+        // TODO: this double casting hack solves the type problem, but is this the right way?
+        callback(this.parseTradeEvent((event as unknown) as Trade))
+      }
+    })
+    // 2: Without, watching only `data` events
+    // subscription.on('data', event => callback(this.parseTradeEvent(event)))
+
+    // TODO: doesn't matter how I subscribe, TS says the type returned by the event does not contain metadata fields, only event fields.
+    // What TS says it returns:
+    // (parameter) event: {
+    //     owner: string;
+    //     orderId: string;
+    //     sellToken: string;
+    //     buyToken: string;
+    //     executedSellAmount: string;
+    //     executedBuyAmount: string;
+    //     0: string;
+    //     1: string;
+    //     2: string;
+    //     3: string;
+    //     4: string;
+    //     5: string;
+    // }
+    // Argument of type '{ owner: string; orderId: string; sellToken: string; buyToken: string; executedSellAmount: string; executedBuyAmount: string; 0: string; 1: string; 2: string; 3: string; 4: string; 5: string; }' is not assignable to parameter of type 'Trade | ContractEventLog<{ owner: string; orderId: string; sellToken: string; buyToken: string; executedSellAmount: string; executedBuyAmount: string; 0: string; 1: string; 2: string; 3: string; 4: string; 5: string; }>'.
+    //   Type '{ owner: string; orderId: string; sellToken: string; buyToken: string; executedSellAmount: string; executedBuyAmount: string; 0: string; 1: string; 2: string; 3: string; 4: string; 5: string; }' is missing the following properties from type 'ContractEventLog<{ owner: string; orderId: string; sellToken: string; buyToken: string; executedSellAmount: string; executedBuyAmount: string; 0: string; 1: string; 2: string; 3: string; 4: string; 5: string; }>': returnValues, event, address, logIndex, and 4 more.ts(2345)
+
+    return (): void => this.unsubscribeToTradeEvent()
+  }
+
+  public unsubscribeToTradeEvent(): void {
+    Object.keys(this.subscriptions.trade).forEach(networkId => {
+      logDebug(`[ExchangeApiImpl] Unsubscribing trade events for network ${networkId}`)
+      this.subscriptions.trade[networkId].unsubscribe()
```

Comment:
Is this fn idempotent?

Reply:
Don't really know what happens if you call unsubscribe on something that is already unsubscribed... I guess nothing?

Reply:
Universe entropy increases

This is where it was supposed to go. Commented through VSCode, didn't work out :disappointed: